src/coreclr/nativeaot/docs/compiling.md

```diff
@@ -153,3 +153,7 @@ Alpine
 ```sh
 apk add numactl-dev
 ```
+
+## NativeAOT on NixOS
+
+NativeAOT relies on hard-coded paths, such as `/lib64/ld-linux-x86-64.so.2`, which do not exist on NixOS. [Here](https://nixos.wiki/wiki/DotNET#NativeAOT) is an instruction for NixOS.
```

Comment:
Where is this dependency coming from? I do not see this path hardcoded anywhere under https://github.com/dotnet/runtime/tree/main/src/coreclr/nativeaot

Reply:
`PT_INTERP` header is included by the linker, by default. We can omit it with, for example:

```csproj
<LinkerArg Include="-Wl,-no-pie -Wl,--no-dynamic-linker" />
```
to get:

```diff
$ file dist/exe1
- dist/exe1: ELF 64-bit LSB pie executable, ARM aarch64, version 1 (SYSV), dynamically linked, interpreter /lib/ld-linux-aarch64.so.1, BuildID[sha1]=715c99294ff6dd56d8df26862f214c884e709046, for GNU/Linux 3.7.0, stripped
+ dist/exe1: ELF 64-bit LSB executable, ARM aarch64, version 1 (SYSV), dynamically linked, BuildID[sha1]=dcf8cfc2034d0374062ce8d8a8bcf93eec053de8, for GNU/Linux 3.7.0, stripped
```

but then the binary fails to execute.

```gdb
Program received signal SIGSEGV, Segmentation fault.
0x00000000000078b0 in ?? ()
(gdb) bt
#0  0x00000000000078b0 in ?? ()
#1  0x0000ffffb2fc8570 in _start ()
```

From ld docs: https://sourceware.org/binutils/docs/ld/Options.html
> --no-dynamic-linker
>
>    When producing an executable file, omit the request for a dynamic linker to be used at load-time. This is only meaningful for ELF executables that contain dynamic relocations, and usually requires entry point code that is capable of processing these relocations

Here is a related article for such custom entrypoint: https://rhysre.net/writing-hello-world-in-c-without-the-dynamic-linker.html. I think we don't want this kind of manual assembly hack. Our pure -static executable doesn't have relocs, but pie-static, pie-dynamic and dynamic executables all have them: `readelf --relocs exe1`

Reply:
@jkotas, I just noticed that our `PositionIndependentExecutable=false` should explicitly set `-Wl,-no-pie` to have an effect because linkers in distros are built with PIE default, and we still get the `pie executable` with `PositionIndependentExecutable=false`. Just omitting `-pie` is not enough. Also the default `-pie` (the compiler option) is unnecessary because we don't compile anything during publish, it should be passed to linker as in: `-Wl,-pie`.

Reply:
Hey @am11 , thanks for the PR, is it possible to change it from within a project by the way? This isn't possible to support all GNU/Linux (let alone non-GNU ones) distros from .NET, so would be nice to have a switch for all these settings like this one :sweat_smile:

Reply:
@WhiteBlackGoose, sorry for the confusion, but that PR was unrelated to dynamic linker requirement. :)

NativeAOT published dynamic executable requires `ld` due to relocs and `-Wl,--no-dynamic-linker` will not help (see the `SIGSEGV` error mentioned above). If we were to support that, we would need to provide our custom `_start` to resolve those relocs. I think it is too much custom work for unknown use-cases.

> is it possible to change it from within a project by the way?

We can add more `LinkerArg` in our project file (.csproj):

```csproj
<ItemGroup>
  <LinkerArg Include="-Wl,..." /> <!-- additional linker option -->
  <LinkerArg Include="path/to/your_object.o" /> <!-- additional object, which can also contain
                                                     your custom _start implementation -->
</ItemGroup>
```

but to remove the linker args added by build integration targets, there is no easy way. I locate the currently used targets files with `dotnet publish -v:diag ... | grep 'Microsoft\.NETCore\.Native.*\.targets'`, make their backup and update them manually (one with `.Unix` in filename has `LinkerArg` entries and one without has `CustomLinkerArg`); and then the usual clean+publish: `rm -rf obj bin && dotnet publish .....`